File: cmd/adminrole.py
import asyncio
import discord
from discord.ext import commands
from discord import app_commands
from utils.permissions.base_admin import BaseAdminCog
from utils.core.embeds import make_embed
from utils.core.emojis import EMOJIS
from db.db_helpers.admin_roles import (add_admin_role, remove_admin_role,
                                       get_admin_roles)


class AdminRole(BaseAdminCog):

    def __init__(self, bot: commands.Bot):
        self.bot = bot

    adminrole = app_commands.Group(
        name="adminrole",
        description=
        "Manage authorized server administration/bot developer roles.",
        default_permissions=discord.Permissions(administrator=True))

    async def _validate(
        self, interaction: discord.Interaction
    ) -> tuple[discord.Guild, discord.Member] | None:
        guild = interaction.guild
        user = interaction.user

        if guild is None or not isinstance(user, discord.Member):
            await interaction.response.send_message(
                f"{EMOJIS.get('error', '❌')} This command can only be executed within a server guild.",
                ephemeral=True)
            return None

        if user.id != guild.owner_id:
            await interaction.response.send_message(
                f"{EMOJIS.get('error', '❌')} Administrative override configurations are restricted strictly to the Server Owner.",
                ephemeral=True)
            return None

        return guild, user

    @adminrole.command(
        name="add",
        description=
        "Authorize a role for administrative bot override permissions.")
    @app_commands.describe(
        role="The target role to grant administrative bot status.")
    async def add(self, interaction: discord.Interaction, role: discord.Role):
        validated = await self._validate(interaction)
        if not validated:
            return
        guild, user = validated

        if role.is_default():
            await interaction.response.send_message(
                f"{EMOJIS.get('warning', '⚠️')} Cannot assign the default `@everyone` profile as an admin role.",
                ephemeral=True)
            return

        if role.managed:
            await interaction.response.send_message(
                f"{EMOJIS.get('warning', '⚠️')} Managed external application or bot integration roles cannot be registered.",
                ephemeral=True)
            return

        bot_member = guild.me
        if bot_member:
            if not bot_member.guild_permissions.manage_roles:
                await interaction.response.send_message(
                    f"{EMOJIS.get('error', '❌')} I lack the `Manage Roles` application permission within this guild hierarchy.",
                    ephemeral=True)
                return

            if role >= bot_member.top_role:
                await interaction.response.send_message(
                    f"{EMOJIS.get('error', '❌')} Action denied: Target role position is equal to or higher than my highest role positioning.",
                    ephemeral=True)
                return

        if role >= user.top_role and user.id != guild.owner_id:
            await interaction.response.send_message(
                f"{EMOJIS.get('error', '❌')} Action denied: You cannot configure roles higher than or equal to your own top status position.",
                ephemeral=True)
            return

        added = await add_admin_role(guild.id, role.id)

        await interaction.response.send_message(embed=make_embed(
            title="Admin Role Configured",
            description=
            (f"{EMOJIS['success']} Authorized {role.mention} as a bot administrator override."
             if added else
             f"{EMOJIS['warning']} Role {role.mention} is already registered in the system."
             ),
            level="SUCCESS" if added else "WARNING",
        ),
                                                ephemeral=True)

    @adminrole.command(
        name="remove",
        description=
        "Revoke administrative bot override authorization from a role.")
    @app_commands.describe(
        role="The target role to strip bot administrative permissions from.")
    async def remove(self, interaction: discord.Interaction,
                     role: discord.Role):
        validated = await self._validate(interaction)
        if not validated:
            return
        guild, _ = validated

        removed = await remove_admin_role(guild.id, role.id)

        await interaction.response.send_message(embed=make_embed(
            title="Admin Role Deauthorized",
            description=
            (f"{EMOJIS['success']} Stripped bot administrative overrides from {role.mention} cleanly."
             if removed else
             f"{EMOJIS['warning']} Target role {role.name} was not flagged as an authorized administrator profile."
             ),
            level="SUCCESS" if removed else "WARNING"),
                                                ephemeral=True)

    @adminrole.command(
        name="list",
        description=
        "Display all custom administrative role definitions for this server.")
    async def list_roles(self, interaction: discord.Interaction):
        validated = await self._validate(interaction)
        if not validated:
            return
        guild, _ = validated

        role_ids = await get_admin_roles(guild.id)
        roles: list[str] = []

        for role_id in role_ids:
            role = guild.get_role(role_id)
            if role:
                roles.append(role.mention)

        description = (
            "\n".join(roles) if roles else
            f"{EMOJIS['warning']} No role configurations discovered within database schemas."
        )

        await interaction.response.send_message(embed=make_embed(
            title="Configured Bot Administrators",
            description=description,
            level="INFO"),
                                                ephemeral=True)

    @adminrole.command(
        name="reset",
        description=
        "Wipe all configured administrative roles from the database.")
    async def reset(self, interaction: discord.Interaction):
        validated = await self._validate(interaction)
        if not validated:
            return
        guild, _ = validated

        role_ids = await get_admin_roles(guild.id)
        if not role_ids:
            await interaction.response.send_message(embed=make_embed(
                title="Registry Already Clear",
                description=
                f"{EMOJIS['warning']} No admin roles are configured for this guild entity.",
                level="WARNING"),
                                                    ephemeral=True)
            return

        await asyncio.gather(*(remove_admin_role(guild.id, role_id)
                               for role_id in role_ids))

        await interaction.response.send_message(embed=make_embed(
            title="Registry Flush Complete",
            description=
            f"{EMOJIS['success']} Flushed all administrative role records cleanly from database caches.",
            level="SUCCESS"),
                                                ephemeral=True)


async def setup(bot: commands.Bot):
    await bot.add_cog(AdminRole(bot))

File: src/commands/admin/adminrole.js
import { SlashCommandBuilder } from "discord.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { EMOJIS } from "../../core/emojis.js";
import {
  addAdminRole,
  getAdminRoles,
  removeAdminRole,
} from "../../db/helpers/adminRoles.js";

const slashBuilder = new SlashCommandBuilder()
  .setName("adminrole")
  .setDescription("Manage authorized bot administration roles (Server Owner only)")
  .addSubcommand((sub) =>
    sub
      .setName("add")
      .setDescription("Authorize a role for administrative bot override permissions")
      .addRoleOption((opt) => opt.setName("role").setDescription("The target role").setRequired(true))
  )
  .addSubcommand((sub) =>
    sub
      .setName("remove")
      .setDescription("Revoke administrative bot authority from a role")
      .addRoleOption((opt) => opt.setName("role").setDescription("The target role").setRequired(true))
  )
  .addSubcommand((sub) =>
    sub.setName("list").setDescription("List all currently authorized admin roles")
  );

export default createCommand({
  name: "adminrole",
  description: "Manage authorized bot administration roles",
  category: "Admin",
  slashBuilder,

  async execute(ctx) {
    const { guild, user, member } = ctx;
    if (!guild) return;

    // Strict Server Owner validation
    if (guild.ownerId !== user.id) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Permission Denied",
            description: `${EMOJIS.get("fail") || "❌"} Administrative role configurations are strictly restricted to the **Server Owner**.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    const sub = ctx.subcommand || ctx.options._args?.[0] || "list";
    const roleId = ctx.options.role || ctx.options._args?.[1]?.replace(/[<@&>]/g, "");

    if (sub === "add") {
      if (!roleId) return await ctx.reply("Please specify a role.");
      const role = guild.roles.cache.get(roleId);
      if (!role) return await ctx.reply("Role not found.");

      const created = await addAdminRole(guild.id, role.id);
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: created ? "Admin Role Added" : "Already Authorized",
            description: created
              ? `${EMOJIS.get("success") || "✅"} Granted bot administration authority to ${role}.`
              : `${EMOJIS.get("warning") || "⚠️"} Role ${role} is already an authorized admin role.`,
            level: created ? "SUCCESS" : "WARNING",
          }),
        ],
      });
    }

    if (sub === "remove") {
      if (!roleId) return await ctx.reply("Please specify a role.");
      const role = guild.roles.cache.get(roleId);
      const deleted = await removeAdminRole(guild.id, roleId);
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: deleted ? "Admin Role Removed" : "Not Authorized",
            description: deleted
              ? `${EMOJIS.get("success") || "✅"} Revoked bot administration authority from ${role ? role : `<@&${roleId}>`}.`
              : `${EMOJIS.get("warning") || "⚠️"} Role is not in the admin role registry.`,
            level: deleted ? "SUCCESS" : "WARNING",
          }),
        ],
      });
    }

    if (sub === "list") {
      const roles = await getAdminRoles(guild.id);
      const roleMentions = roles
        .map((rId) => {
          const r = guild.roles.cache.get(rId);
          return r ? `• ${r} (\`${r.id}\`)` : `• <@&${rId}> (\`${rId}\`)`;
        })
        .join("\n");

      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: `Admin Roles • ${guild.name}`,
            description:
              roles.length > 0
                ? `Authorized roles with bot admin permissions:\n\n${roleMentions}`
                : `${EMOJIS.get("warning") || "⚠️"} No admin roles configured. Only Server Owner & Administrators have authority.`,
            level: "INFO",
          }),
        ],
      });
    }
  },
});
